x-pack/plugins/cases/server/connectors/cases/cases_oracle_service.test.ts

```diff
@@ -0,0 +1,283 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { createHash } from 'node:crypto';
+import stringify from 'json-stable-stringify';
+import { savedObjectsClientMock } from '@kbn/core-saved-objects-api-server-mocks';
+import { loggerMock } from '@kbn/logging-mocks';
+
+import { CasesOracleService } from './cases_oracle_service';
+import { CASE_ORACLE_SAVED_OBJECT } from '../../../common/constants';
+import { isEmpty, set } from 'lodash';
+
+describe('CasesOracleService', () => {
+  const unsecuredSavedObjectsClient = savedObjectsClientMock.create();
+  const mockLogger = loggerMock.create();
+
+  let service: CasesOracleService;
+
+  beforeEach(() => {
+    jest.resetAllMocks();
+    service = new CasesOracleService({ unsecuredSavedObjectsClient, log: mockLogger });
+  });
+
+  describe('getRecordId', () => {
+    it('return the record ID correctly', async () => {
+      const ruleId = 'test-rule-id';
+      const spaceId = 'default';
+      const owner = 'cases';
+      const grouping = { 'host.ip': '0.0.0.1' };
+
+      const payload = `${ruleId}:${spaceId}:${owner}:${stringify(grouping)}`;
+      const hash = createHash('sha256');
+
+      hash.update(payload);
+
+      const hex = hash.digest('hex');
+
+      expect(service.getRecordId({ ruleId, spaceId, owner, grouping })).toEqual(hex);
+    });
+
+    it('sorts the grouping definition correctly', async () => {
+      const ruleId = 'test-rule-id';
+      const spaceId = 'default';
+      const owner = 'cases';
+      const grouping = { 'host.ip': '0.0.0.1', 'agent.id': '8a4f500d' };
+      const sortedGrouping = { 'agent.id': '8a4f500d', 'host.ip': '0.0.0.1' };
```

Comment:
Where is this done in the implementation?

Reply:
Here https://github.com/elastic/kibana/pull/168370/files/c210a0c9cefc167ac1c1339757ec8494729abb4e#diff-6d91ad0215d2086dc33208df409ca2e693725d41e076a04b4bee371b36cfabf6R24. `json-stable-stringify` always return a sorted version of the object. We need the order to be the same to produce the same hash for the same grouping no matter how the object is defined (order is not guaranteed by JS neither we should care in the application level).